src/server/bg_services/replication_log_parser.js

```diff
@@ -63,6 +66,45 @@ async function get_aws_log_candidates(source_bucket_id, rule_id, replication_con
     };
 }
 
+/**
+ * analyze_candidates will iterate over all the candidates and will return an array of candidates
+ * such that there will be ONLY one action per object. It will do so by sorting the candidates
+ * by time and in case of conflicts, it will replace the action with 'conflict'. The function
+ * that consumes the candidates should handle the conflict action.
+ * @param {nb.ReplicationLogCandidates} candidates 
+ * @returns {nb.ReplicationLogCandidates}
+ */
+function analyze_candidates(candidates) {
+    /**
+     * @type {Map<Date, Record<string, nb.ReplicationLogCandidate>>}
+     */
+    const candidates_map = new Map();
+
+    // group the candidates by time - there should only 1 candidate per time per key
+    // in case of a conflict, we will mark the candidate with 'conflict' action
+    for (const candidate of candidates) {
+        const { key, time } = candidate;
+        if (!candidates_map.has(time)) {
+            candidates_map.set(time, {});
+        }
+
+        const time_map = candidates_map.get(time);
+
+        // if there is already a candidate for the same key, then we have a conflict
+        if (time_map[key]) {
+            const conflict_candidate = time_map[key];
+            conflict_candidate.action = 'conflict';
```

Comment:
should we avoid here versioning enabled buckets? this "squash" will not create the same versions as we have in the source bucket

Reply:
Yes, I had questions about versioned objects when I picked up the task. Unfortunately, we don't yet have a "doc" yet (not that I know of) which clarifies how we deal with the cases like:
- Source has versioning destination doesn't.
- Destination has versioning source doesn't.
- Both have versioning.

eg. What to do with delete markers when destination doesn't have versioning? Not that it can't be done, just that we don't have it documented yet.

This might be a broader discussion I think, that's why the implementation currently does not do anything special when it comes to versioned objects.

Reply:
Alright, for creating a better user experience I think that we should document it